pyasn1/codec/ber/decoder.py

```diff
@@ -4,45 +4,117 @@
 # Copyright (c) 2005-2019, Ilya Etingof <etingof@gmail.com>
 # License: http://snmplabs.com/pyasn1/license.html
 #
+import os
+import sys
+from io import BytesIO, BufferedReader
+
 from pyasn1 import debug
 from pyasn1 import error
 from pyasn1.codec.ber import eoo
 from pyasn1.compat.integer import from_bytes
 from pyasn1.compat.octets import oct2int, octs2ints, ints2octs, null
+from pyasn1.error import PyAsn1Error
 from pyasn1.type import base
 from pyasn1.type import char
 from pyasn1.type import tag
 from pyasn1.type import tagmap
 from pyasn1.type import univ
 from pyasn1.type import useful
 
-__all__ = ['decode']
+
+__all__ = ['decodeStream']
 
 LOG = debug.registerLoggee(__name__, flags=debug.DEBUG_DECODER)
 
 noValue = base.noValue
 
 
+_BUFFER_SIZE = 1024
+_PY2 = sys.version_info < (3,)
+
+
+def asSeekableStream(substrate):
+    """Convert object to seekable bytes stream.
+
+    :type substrate: Union[bytes, IOBase, univ.OctetString]
+    :rtype: IOBase
+    """
+    if isinstance(substrate, bytes):
+        return BytesIO(substrate)
+    elif isinstance(substrate, univ.OctetString):
+        return BytesIO(substrate.asOctets())
+    try:
+        if _PY2 and isinstance(substrate, file):
+            return BytesIO(substrate.read())  # Not optimal for really large files
+        elif not substrate.seekable():
+            return BufferedReader(substrate, _BUFFER_SIZE)
+        else:
+            return substrate
+    except AttributeError as f:
+        print(f)
+        raise TypeError("Cannot convert " + substrate.__class__.__name__ + " to seekable bit stream.")
+
+
+def endOfStream(substrate):
+    """Check whether we have reached an end of stream.
+
+    :type substrate: IOBase
+    :rtype: bool
+    """
+    if isinstance(substrate, BytesIO):
+        cp = substrate.tell()
+        substrate.seek(0, os.SEEK_END)
```

Comment:
Seeking to the end of a large file can be expensive... Can we seek by just one byte?

Reply:
Seeking to the end of stream is used only for BytesIO (it just moves some pointer to a position that is inaccessible through public API). This is the fastest check of end-of-stream I was able to implement.

With non-seekable streams, I have a problem. It is not possible to back-track in them (and BufferedReader does not solve it - my assumption in this was incorrect). One possible solution to this is to create a wrapper around them that caches a couple of last bytes read. What do you think?

Reply:
> Seeking to the end of stream is used only for BytesIO (it just moves some pointer to a position that is inaccessible through public API). This is the fastest check of end-of-stream I was able to implement.

Makes sense to me!

> With non-seekable streams, I have a problem. It is not possible to back-track in them (and BufferedReader does not solve it - my assumption in this was incorrect). One possible solution to this is to create a wrapper around them that caches a couple of last bytes read. What do you think?

I had exactly the same though! I suspected that there may be non-seekable streams that we still might want to handle (e.g. network connection).

How deeply do we need to backtrack?

Reply:
> How deeply do we need to backtrack?

I don't think we need to get back more than a couple of bytes (the tag + the length). It is only necessary to reinterpret the data in AnyDecoder - at least this is the only place with explicit `seek`.

Reply:
Perhaps, however, I suspect that formally, both tag and length can be quite (indefinitely?) large... ;)

Reply:
> With non-seekable streams, I have a problem. It is not possible to back-track in them (and BufferedReader does not solve it - my assumption in this was incorrect). One possible solution to this is to create a wrapper around them that caches a couple of last bytes read. What do you think?

I've come across an attempt to solve distantly similar problem in the [gzip-as-a-file](https://github.com/python/cpython/blob/2.7/Lib/gzip.py#L337) implementation. That one  appears quite limited. Mentioning just for inspiration. o/

Reply:
Hopefully solved with the wrapper in e27f97182e859f.